cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -0,0 +1,289 @@
+ * limitations under the License.
+ */
+
+#include <cmath>
+#include <limits>
+#include <functional>
+#include <boost/serialization/strong_typedef.hpp>
+
+namespace cudf {
+namespace fp {
+
+BOOST_STRONG_TYPEDEF(int32_t, scale_type)
+
+enum Radix : int32_t {
+    BASE_2  = 2,
+    BASE_10 = 10
+};
+
+namespace detail {
+    // helper function to negate strongly typed scale_type
+    auto negate(scale_type const& scale) -> scale_type{
+        return scale_type{-1 * scale};
+    }
+
+    // perform this operation when constructing with - scale (after negating scale)
+    template <Radix Rad, typename T>
+    constexpr auto right_shift(T const& val, scale_type const& scale) {
+        return val / std::pow(static_cast<int32_t>(Rad), static_cast<int32_t>(scale));
+    }
+
+    // perform this operation when constructing with + scale
+    template <Radix Rad, typename T>
+    constexpr auto left_shift(T const& val, scale_type const& scale) {
+        return val * std::pow(static_cast<int32_t>(Rad), static_cast<int32_t>(scale));
+    }
+
+    // convenience generic shift function
+    template <Radix Rad, typename T>
+    constexpr auto shift(T const& val, scale_type const& scale) {
+        return scale >= 0 ? right_shift<Rad>(val, scale)
+                          : left_shift <Rad>(val, negate(scale));
+    }
+}
+
+// helper struct for constructing fixed_point when value is already shifted
+template <typename Rep>
+struct scaled_integer{
+    Rep value;
+    scale_type scale;
+    explicit scaled_integer(Rep v, scale_type s) : value(v), scale(s) {}
+};
+
+
+// Rep = representative type
+template <typename Rep, Radix Rad>
+class fixed_point {
+
+    scale_type _scale;
+    Rep        _value;
+
+public:
+
+    // CONSTRUCTORS
+    template <typename T = Rep,
+              typename std::enable_if_t<(std::numeric_limits<T>::is_integer
+                                      || std::is_floating_point<T>::value)>* = nullptr>
+    explicit fixed_point(T const& value, scale_type const& scale) :
+        _value(detail::shift<Rad>(value, scale)),
+        _scale(scale)
+    {
+    }
+
+    explicit fixed_point(scaled_integer<Rep> s) :
+        _value(s.value),
+        _scale(s.scale)
+    {
+    }
+
+    // EXPLICIT CONVERSION OPERATOR
+    template <typename U,
+              typename std::enable_if_t<(std::numeric_limits<U>::is_integer
+                                      || std::is_floating_point<U>::value)>* = nullptr>
+    explicit constexpr operator U() const {
+        return detail::shift<Rad>(static_cast<U>(_value), detail::negate(_scale));
+    }
+
+    auto get() const noexcept {
+        auto const underlying_scale     = static_cast<int32_t>(_scale);
+        int  const rounded_val          = _value / std::pow(static_cast<int32_t>(Rad), underlying_scale);
+        bool const needs_floating_point = rounded_val * Rad * underlying_scale != _value;
+        return needs_floating_point ? static_cast<double>(*this) : static_cast<Rep>(*this);
```

Comment:
The key is line 102 where we are doing integer division: `int  const rounded_val` meaning that when decimal points aren't actually needed, we will take the `static_cast<Rep>` path. It might be clearer if instead of `Rep` I just used `int32_t`. What do you think?

Reply:
I think it makes sense when the `static_cast<Rep>` path is taken, but I'm thinking that the return type of the function must always be `double` since all paths must return the same type. So even if the `static_cast<Rep>` path is taken and `Rep` is an integer, won't the result still be cast to the return type `double`?

Reply:
Wow, I don't know how I missed this. You are absolutely correct. Great catch.
https://godbolt.org/z/D-LGBD
```cpp
auto deduced_return_type(bool cond) {
    return cond ? 4 : 4.2;
}

auto main() -> int { 

    std::cout << deduced_return_type(false) << '\n'; // 4.2
    std::cout << deduced_return_type(true)  << '\n'; // 4

    std::cout << std::is_same_v<decltype(deduced_return_type(false)), double> << '\n'; // 1
    std::cout << std::is_same_v<decltype(deduced_return_type(true)),  double> << '\n'; // 1

    return 0; 
}
```
Let me clean this up then.

Reply:
Done.